File: examples/CSharp/HelloWorld/CustomSettingsExample.cs
using System;
using System.Threading.Tasks;

using Microsoft.Extensions.Configuration;

using NBomber.Contracts;
using NBomber.CSharp;

namespace CSharp.HelloWorld
{
    public class CustomScenarioSettings
    {
        public int TestField { get; set; }
        public int PauseMs { get; set; }
    }

    public class CustomSettingsExample
    {
        static CustomScenarioSettings _customSettings = new CustomScenarioSettings();

        static Task TestInit(IScenarioContext context)
        {
            _customSettings = context.CustomSettings.Get<CustomScenarioSettings>();

            context.Logger.Information(
                "test init received CustomSettings.TestField '{TestField}'",
                _customSettings.TestField
            );

            return Task.CompletedTask;
        }

        public static void Run()
        {
            var step = Step.Create("step", async context =>
            {
                await Task.Delay(TimeSpan.FromSeconds(0.1));

                context.Logger.Debug(
                    "step received CustomSettings.TestField '{TestField}'",
                    _customSettings.TestField
                );

                return Response.Ok(); // this value will be passed as response for the next step
            });

            var customPause = Step.CreatePause(() => _customSettings.PauseMs);

            var scenario = ScenarioBuilder
                .CreateScenario("my_scenario", step, customPause)
                .WithoutWarmUp()
                .WithInit(TestInit)
                .WithLoadSimulations(new []
                {
                    Simulation.KeepConstant(copies: 1, TimeSpan.FromSeconds(30))
                });

            NBomberRunner
                .RegisterScenarios(scenario)
                .WithTestSuite("example")
                .WithTestName("custom_settings")
                .LoadConfig("./HelloWorld/config.json")
                .Run();
        }
    }
}

File: examples/FSharp/HelloWorld/CustomSettingsExample.fs
module FSharp.HelloWorld.CustomSettingsExample

open System.Threading.Tasks

open Microsoft.Extensions.Configuration
open FSharp.Control.Tasks.V2.ContextInsensitive

open NBomber.Contracts
open NBomber.FSharp

[<CLIMutable>]
type CustomScenarioSettings = {
    TestField: int
    PauseMs: int
}

let run () =

    let mutable _customSettings = { TestField = 0; PauseMs = 0 }

    let scenarioInit (context: IScenarioContext) = task {
        _customSettings <- context.CustomSettings.Get<CustomScenarioSettings>()

        context.Logger.Information(
            "test init received CustomSettings.TestField '{TestField}'",
            _customSettings.TestField
        )
    }

    let step = Step.create("step", fun context -> task {

        do! Task.Delay(seconds 1)

        context.Logger.Debug(
            "step received CustomSettings.TestField '{TestField}'",
            _customSettings.TestField
        )

        return Response.Ok()
    })

    let customPause = Step.createPause(fun () -> _customSettings.PauseMs)

    Scenario.create "my_scenario" [step; customPause]
    |> Scenario.withInit scenarioInit
    |> NBomberRunner.registerScenario
    |> NBomberRunner.loadConfig "./HelloWorld/config.json"
    |> NBomberRunner.run
    |> ignore
